## Installing required packages

In [105]:
!pip install altair --upgrade --quiet

## Importing libraries

In [1]:
import numpy as np
import pandas as pd
import altair as alt
from altair.expr import datum
import ast
alt.data_transformers.disable_max_rows();

## The Movies Dataset from kaggle
#### https://www.kaggle.com/rounakbanik/the-movies-dataset
##### movies_metadata.csv, credits.csv

In [21]:
# credits.csv github file path
credits_url = 'https://raw.githubusercontent.com/ansjayan/project_jovianml/master/credits.csv.zip'
# credits_url = '/home/ans/Documents/jovianml/project_jovianml/credits.csv.zip'

# movies_metadata.csv github file path
movies_url = 'https://raw.githubusercontent.com/ansjayan/project_jovianml/master/movies_metadata.csv.zip'
# movies_url = '/home/ans/Documents/jovianml/project_jovianml/movies_metadata.csv.zip'
# reading movies_metadata.csv
movies = pd.read_csv(movies_url, low_memory = False)
# reading credits.csv
credits = pd.read_csv(credits_url)

In [22]:
# first 5 rows of movies dataframe
movies.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [23]:
# summary of movies dataframe
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [24]:
# first 5 rows of credits dataframe
credits.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [25]:
# summary of credits dataframe
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB


In [7]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

NameError: name 'jovian' is not defined

# Data Preparation and Cleaning

In [8]:
# number of na in movies id, isna returns boolean series, which is summed
print(movies.id.isna().sum())
# number of na in movies title, isna gives boolean series, which is summed
print(movies.title.isna().sum())

0
6


In [9]:
# filling na in title column with values form original_title
# if title has na, that is replaced with value from original_title
movies.title = movies.title.fillna(value = movies.original_title)
movies.title.isna().sum()

0

In [10]:
# title and id is required for analysis, if na in id column remove that rows
movies.dropna(subset = ["id"], inplace = True)
# number of na in id column, isna gives boolean series, which is summed
movies.id.isna().sum()

0

In [11]:
# converting movies id to int from float
# movies.id = movies.id.astype("int") # other equivalent is pd.to_numeric()
# to_numeric converts id to numeric values
movies["id"] = pd.to_numeric(movies.id, errors = 'coerce', downcast = "integer")

In [12]:
# Is there any movies id not in credits id.
# isin returns boolean values, ~ negated
movies[~movies.id.isin(credits.id)]

movies_metadata.csv belongs_to_collection budget genres homepage  \
42883               False                   NaN      0     []      NaN   

             id    imdb_id original_language original_title  \
42883  401840.0  tt3291632                es   School's out   

                                                overview  ... release_date  \
42883  Two high school kids mentored by a nightclub o...  ...   2017-05-30   

      revenue runtime spoken_languages    status  tagline         title  \
42883     0.0     0.0               []  Released      NaN  School's out   

       video vote_average vote_count  
42883  False          0.0        0.0  

[1 rows x 24 columns]

In [13]:
# Is there any credits id not in movies id
credits[~credits.id.isin(movies.id)]

Empty DataFrame
Columns: [credits.csv, crew, id]
Index: []

In [14]:
# Merging movies to credits.
# as credits has only 3 columns which are useless if others[movies] are not there.
# merging on id column, using left method. here movies is left.
movies = movies.merge(credits, how = "left", left_on = "id", right_on = "id")

In [15]:
# return counts of unique values.
movies.status.value_counts()

Released           45088
Rumored              232
Post Production       98
In Production         20
Planned               15
Canceled               2
Name: status, dtype: int64

In [16]:
# Selecting only released movies
movies = movies.loc[movies.status == 'Released'].copy()

In [17]:
# unique values of id column
# print(movies.id.value_counts())
# checking duplicated columns
# print(movies[movies.duplicated(subset = ['id'], keep = False)].sort_values(by = 'id').head())
# dropping duplicated values
movies.drop_duplicates(inplace = True)
# dropping duplicated values based on id column
movies.drop_duplicates(subset = "id", inplace = True)

In [18]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

NameError: name 'jovian' is not defined

 ## Converting columns to proper format

In [19]:
# converting adult column dtype object to bool.
# apply method applies a lambda function, which replace str to bool.
movies.adult = movies.adult.apply(lambda x: False if x == 'False' else True)

AttributeError: 'DataFrame' object has no attribute 'adult'

In [125]:
# ['belongs_to_collection'] contains stringified dict. 
# ['genres', 'production_countries', 'production_companies', 'spoken_languages'] all contains stringified list of dict/json.
json_col = ['belongs_to_collection', 'genres', 'production_countries', 'production_companies', 'spoken_languages', 'cast', 'crew']

In [126]:
# ['belongs_to_collection'] contains stringified dict.
movies.belongs_to_collection[0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

In [127]:
# converts stringified dict to dict
ast.literal_eval(movies.belongs_to_collection[0])

{'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg',
 'id': 10194,
 'name': 'Toy Story Collection',
 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg'}

In [128]:
# functions to convert columns from str to list/json/dict
def convert_if_str(column):
    movies[column] = movies[column].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [129]:
# Converting columns from str to list/json/dict
for i in json_col:
    convert_if_str(i)

In [130]:
# creating new column for total number of cast
movies["cast_size"] = movies.cast.apply(lambda x: np.nan if isinstance(x, float) else len(x))
movies["crew_size"] = movies.crew.apply(lambda x: np.nan if isinstance(x, float) else len(x))
# certain values are nan-float and others are list.
# len don't work for nan, so nan retained as nan by checking isinstance.
# len of list is calculated by len(x).

In [131]:
movies.belongs_to_collection[0]['name']# if isinstance(x, dict) else np.nan)

'Toy Story Collection'

In [132]:
# functions to extracting required values from columns.
def extract_if_dict(column, value):
    movies[column] = movies[column].apply(lambda x: x[value] if isinstance(x, dict) else np.nan)
    movies[column].replace("", np.nan, inplace = True)    

def extract_if_list_dict(column, value):
    movies[column] = movies[column].apply(lambda x: "|".join(i[value] for i in x) if isinstance(x, list) else np.nan)
    movies[column].replace("", np.nan, inplace = True)    

In [133]:
# Extracting required values form columns
# ['genres', 'production_countries', 'production_companies', 'spoken_languages'] all contains list of dict/json.
for i in json_col[1:-1]:
    extract_if_list_dict(i, "name")
    
# ['belongs_to_collection'] belong to dict.     
for i in json_col[0:1]:
    extract_if_dict(i, "name")

In [134]:
# function to extract director
def director(x):
    if isinstance(x, float):
        return np.nan
    else:
        for i in x:
            if i['job'] == 'Director':
                return i['name']
#               else: return np.nan # will change for subsequent jobs to np.nan , makes error
        return np.nan

# function to extract screenplay
def screenplay(x):
    if isinstance(x, float):
        return np.nan
    else:
        for i in x:
            if i['job'] == 'Screenplay':
                return i['name']
#               else: return np.nan # will change for subsequent jobs to np.nan , makes error
        return np.nan


In [135]:
# creating new director, screenplay columns
movies['director'] = movies.crew.apply(director)
movies['screenplay'] = movies.crew.apply(screenplay)

In [136]:
# ['budget', 'revenue', 'runtime', 'popularity', 'vote_count', 'vote_average'] convert to numeric type from object.
numeric_cols = ['budget', 'revenue', 'runtime', 'popularity', 'vote_count', 'vote_average']
# function to convert object columns to numeric types
def convert_to_numeric(column):
    movies[column] = pd.to_numeric(movies[column], errors = 'coerce')
    movies[column] = movies[column].replace(0, np.nan)
# converting object columns to numeric types
for i in numeric_cols:
    convert_to_numeric(i)    

In [137]:
# Budget and revenue in millions USD
movies.budget = movies.budget.div(1000000)
movies.revenue = movies.revenue.div(1000000)
# Adding profit and returns columns, both are in million USD
movies['profit'] = movies.revenue.sub(movies.budget)
movies['returns'] = movies.revenue.div(movies.budget)

In [138]:
# Release_date in datetime format
movies.release_date = pd.to_datetime(movies.release_date, errors='coerce')

In [139]:
# movies.original_language, no cleaning required
movies.original_language.value_counts(dropna = False).head()

en    31935
fr     2416
it     1511
ja     1339
de     1073
Name: original_language, dtype: int64

In [140]:
# overview column requires cleaning to remove No review rows
# movies.overview.value_counts(dropna=False)
movies.overview.replace(["No overview found.", "No Overview", "No movie overview available.", "-", "", "No overview yet.", " "], np.nan, inplace = True)

In [141]:
# tagline column has some space, - which must be removed
# movies.tagline.value_counts(dropna=False)
movies.tagline.replace(["", " ", "-"], np.nan, inplace = True)

In [142]:
jovian.commit(project='my-project', environment=None)

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Committed successfully! https://jovian.ml/ansjayan/my-project


'https://jovian.ml/ansjayan/my-project'

## Dropping na

In [143]:
# checking notna in rows
# movies.notna().sum(axis=1).value_counts().sort_index(ascending = False)

In [144]:
# approx 30 columns
# na 10 or above is removed
# keep rows with atleast 9 nonna values
movies.dropna(thresh = 9, inplace = True) 

In [145]:
col = ['budget', 'revenue', 'vote_count', 'vote_average', 'popularity', 'runtime', 'cast_size', 'crew_size', 'profit', 'returns']
# rounding float values to decimal places
for value in col:
    movies[col] = movies[col].apply(lambda x:round(x, ndigits=2))
    movies[col] = movies[col].replace(0, np.nan)

In [146]:
jovian.commit(project='my-project', environment=None)

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Committed successfully! https://jovian.ml/ansjayan/my-project


'https://jovian.ml/ansjayan/my-project'

## Dropping some columns unnecessary for this project

In [147]:
# dropping columns
movies.drop(columns = ['adult', 'homepage', 'imdb_id', 'original_title', 'poster_path', 'status', 'video', 'crew'], inplace = True)
# adult = adult content or not
# homepage = homepage details, most of them are missing nan
# imdb_id = imdb id is not required as tmdb id column is there
# original title = not required as english title is there
# poster_path = gives url paths for poster
# status = only released is selected in the begining.
# video = trailer is there or not
# crew = extracted director and screenplay from it.

In [148]:
# list to rearrange columns in movies
col = ["id", "title", "director", "screenplay", "tagline", "release_date", "genres", "belongs_to_collection", 
       "original_language", "budget", "revenue", "production_companies",
       "production_countries", "vote_count", "vote_average", "popularity", "runtime",
       "overview", "spoken_languages", 'cast', 'cast_size', 'crew_size', 'profit', 'returns']

In [149]:
# rearrangin columns as in col list
movies = movies.loc[:, col]

In [150]:
# resetting and droping index
movies.reset_index(drop = True, inplace = True)

In [151]:
# write to movies_final csv
movies.to_csv("movies_final.csv", index = False)

In [152]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Committed successfully! https://jovian.ml/ansjayan/my-project


'https://jovian.ml/ansjayan/my-project'

# The Movies Dataset
#### "movies_final.csv"
* __*id:*__ The ID of the movie, TMDB.
* **_title:_** The Official Title of the movie.
* **_director:_** The Director of the movie.
* **_screenplay:_** Screenplay writer of the movie.
* **_tagline:_** The tagline of the movie.
* **_release_date:_** The Theatrical Release Date of the movie.
* **_genres:_** The Genres associated with the movie.
* **_belongs_to_collection:_** Information on the movie series, the particular film belongs to.
* **_original_language:_** The language in which the movie was originally shot in.
* **_budget:_** The budget of the movie in million dollars.
* **_revenue:_** The total revenue of the movie in million dollars.
* **_production_companies:_** The Production companies involved in the making of the movie.
* **_production_countries:_** The Countries where the movie was shot in.
* **_vote_count:_** The number of votes by users, as counted by TMDB.
* **_vote_average:_** The average rating of the movie.
* **_popularity:_** The Popularity Score assigned by TMDB.
* **_runtime:_** The runtime of the movie in minutes.
* **_spoken_languages:_** The Spoken languages in the film.
* **_cast:_** The Actors appearing in the movie.
* **_cast_size:_** The number of Actors appearing in the movie.
* **_crew_size:_** The Size of the film crew, excluding actors.
* **_profit:_** The profit im millions usd for each movie.
* **_returns:_** The returns im millions usd for each movie.

# Exploratory data analysis

In [153]:
# reading movies_final csv to movies dataframe
movies = pd.read_csv("movies_final.csv", parse_dates=['release_date'])

In [154]:
# first 5 rows
movies.head()

id                        title  ...  profit returns
0    862.0                    Toy Story  ...  343.55   12.45
1   8844.0                      Jumanji  ...  197.80    4.04
2  15602.0             Grumpier Old Men  ...     NaN     NaN
3  31357.0            Waiting to Exhale  ...   65.45    5.09
4  11862.0  Father of the Bride Part II  ...     NaN     NaN

[5 rows x 24 columns]

In [155]:
# summary info of movies
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44985 entries, 0 to 44984
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     44985 non-null  float64       
 1   title                  44985 non-null  object        
 2   director               44125 non-null  object        
 3   screenplay             15635 non-null  object        
 4   tagline                20285 non-null  object        
 5   release_date           44907 non-null  datetime64[ns]
 6   genres                 42601 non-null  object        
 7   belongs_to_collection  4463 non-null   object        
 8   original_language      44975 non-null  object        
 9   budget                 8576 non-null   float64       
 10  revenue                7198 non-null   float64       
 11  production_companies   33357 non-null  object        
 12  production_countries   38839 non-null  object        
 13  v

In [156]:
# summarizing numerical columns
movies.describe()

id       budget  ...       profit       returns
count   44985.000000  8576.000000  ...  5309.000000  5.309000e+03
mean   108100.065689    22.372351  ...    60.019260  5.641379e+03
std    112318.916473    34.686531  ...   140.308394  2.183215e+05
min         2.000000     0.010000  ...  -165.710000  1.000000e-02
25%     26225.000000     2.300000  ...    -1.630000  8.300000e-01
50%     59820.000000     9.000000  ...    11.590000  2.090000e+00
75%    157353.000000    26.000000  ...    63.930000  4.280000e+00
max    469172.000000   380.000000  ...  2550.970000  1.239638e+07

[8 rows x 11 columns]

In [157]:
# summarizing categorical columns
movies.describe(include = "object")

title   director  ... spoken_languages            cast
count        44985      44125  ...            41110           42625
unique       41888      17425  ...             1828           42287
top     Cinderella  John Ford  ...          English  Georges Méliès
freq            11         66  ...            22190              24

[4 rows x 12 columns]

In [158]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Committed successfully! https://jovian.ml/ansjayan/my-project


'https://jovian.ml/ansjayan/my-project'

## Director

In [159]:
# Director dataframe
director = movies.loc[:, ['title', 'director', 'revenue', 'budget', 'profit', 'returns']]

### Who directed maximum number of movies?. 

In [160]:
# Maximum number of movies directed by.
dir = director.director.value_counts().reset_index()
dir.columns = ['director', 'counts']
# creating an altair bar chart
bars = alt.Chart(dir.head(20)).mark_bar().encode(
    x = 'director:O',
    y = 'counts:Q',
    color = alt.condition(
        datum.counts == np.amax(dir.counts),
        alt.value('red'),
        alt.value('green')
    ),
    tooltip = 'director'
).properties(
    title='Most Number of Films Directed by',
    height=200,
    width=500
)
# adding text info
text = bars.mark_text(
    align='center', # center left right
    baseline='bottom', # top middle
    dx=0  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    text='counts:Q'
)
# layering both bars and text
alt.layer(bars, text)


alt.LayerChart(...)

### Total profits attained from all the movies of each director?.   


In [161]:
# directors with maximum profit
profit = director.groupby('director').profit.sum().sort_values(ascending=False).reset_index()
profit.columns = ['director', 'profit']
profit['profit'] = profit['profit'].astype("int")

bars = alt.Chart(profit.head(20)).mark_square().encode(
    x='director:O',
    y='profit:Q',
    color = alt.condition(
        datum.profit == np.amax(profit['profit']),
        alt.value('red'),
        alt.value('green')
    ),
    tooltip='director'
).properties(
    title='Directors with maximum profit',
    height=200,
    width=500
)
text = bars.mark_text(
    align='center', # center left right
    baseline='bottom', # top middle
    dx=0  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    text='profit:Q'
)
alt.layer(bars, text)


alt.LayerChart(...)

### Which director's movie got maximum profit?

In [162]:
# Highest profit movies.
hpt = director.sort_values(by='profit', ascending=False)
hpt.dropna(inplace=True)
hpt.profit = hpt.profit.astype("int")

points = alt.Chart(hpt.head(20)).mark_tick().encode(
    x='director:O',
    y='profit:Q',
    color = alt.condition(
        datum.profit == np.amax(hpt['profit']),
        alt.value('red'),
        alt.value('green')
    ),
    tooltip=['title', 'director', 'profit'] # shows movies name
).properties(
    title='Highest profit',
    height=200,
    width=500
)
text = points.mark_text(
    align='center', # center left right
    baseline='bottom', # top middle
    dx=0  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    text='profit:Q',
)
alt.layer(points, text)

alt.LayerChart(...)

In [163]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
[jovian] Committed successfully! https://jovian.ml/ansjayan/my-project


'https://jovian.ml/ansjayan/my-project'

## Screenplay Writers

In [164]:
# screenplay dataframe
screenplay = movies.loc[:, ['title', 'screenplay', 'revenue', 'budget', 'profit', 'returns']]

### Who wrote maximum number of movies?. 

In [165]:
# Maximum number of movies written by.
scr = screenplay.screenplay.value_counts().reset_index()
scr.columns = ['screenplay', 'counts']
# creating an altair bar chart
bars = alt.Chart(scr.head(20)).mark_rule().encode(
    x = 'screenplay:O',
    y = 'counts:Q',
    color = alt.condition(
        datum.counts == np.amax(scr.counts),
        alt.value('red'),
        alt.value('green')
    ),
    tooltip='screenplay'
).properties(
    title='Most Number of Films Written by',
    height=200,
    width=500
)
# adding text info
text = bars.mark_text(
    align='center', # center left right
    baseline='bottom', # top middle
    dx=0  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    text='counts:Q'
)
# layering both bars and text
alt.layer(bars, text)


alt.LayerChart(...)

### Total profits attained from all the movies of each writer?.   


In [166]:
# writers with maximum profit
profit = screenplay.groupby('screenplay').profit.sum().sort_values(ascending=False).reset_index()
profit.columns = ['screenplay', 'profit']
profit['profit'] = profit['profit'].astype("int")

bars = alt.Chart(profit.head(20)).mark_bar().encode(
    x='screenplay:O',
    y='profit:Q',
    color = alt.condition(
        datum.profit == np.amax(profit['profit']),
        alt.value('red'),
        alt.value('green')
    ),
    tooltip='screenplay'
).properties(
    title='Screenplay writers with maximum profit',
    height=200,
    width=500
)
text = bars.mark_text(
    align='center', # center left right
    baseline='bottom', # top middle
    dx=0  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    text='profit:Q'
)
alt.layer(bars, text)


alt.LayerChart(...)

### Which screenplay writer's movie got maximum profit?

In [167]:
# Highest profit movies.
hpt = screenplay.sort_values(by='profit', ascending=False)
hpt.dropna(inplace=True)
hpt.profit = hpt.profit.astype("int")

points = alt.Chart(hpt.head(20)).mark_text().encode(
    x='screenplay:O',
    y='profit:Q',
    color = alt.condition(
        datum.profit == np.amax(hpt['profit']),
        alt.value('red'),
        alt.value('green')
    ),
    tooltip=['title', 'screenplay', 'profit'] # shows movies name
).properties(
    title='Highest profit',
    height=200,
    width=500
)
text = points.mark_text(
    align='center', # center left right
    baseline='bottom', # top middle
    dx=0  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    text='profit:Q',
)
alt.layer(points, text)

alt.LayerChart(...)

In [ ]:
jovian.commit(project='my-project', environment=None)

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...


## Release Year 

In [ ]:
ry = movies.loc[:, ['release_date', 'title', 'revenue', 'budget']]
ry.release_date.dropna(inplace=True)
ry['release_year'] = ry.release_date.apply(lambda x:x.year)
ry.drop(columns=['release_date'], inplace=True)
ry = ry.convert_dtypes()

### Total number of movies released every year, highest of it?. 

In [ ]:
# Total number of movies released every year, highest of it?.
yt = ry.release_year.value_counts().sort_values(ascending=False).reset_index()
yt.columns = ['year', 'counts']

bars = alt.Chart(yt.head(20)).mark_bar().encode(
    x='year:O',
    y='counts:Q',
    color = alt.condition(
        datum.counts == np.amax(yt['counts']),
        alt.value('red'),
        alt.value('green')
    ),
    tooltip = ['counts', 'year']
).properties(
    title='Which Year Maximum movies released.',
    height=200,
    width=500
)
text = bars.mark_text(
    align='center', # center left right
    baseline='bottom', # top middle
    dx=0  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    text='counts:Q'
)
alt.layer(bars, text)



### How budget varies every year?.

In [ ]:
by = ry.sort_values(by='release_year', ascending=False)
by = by.loc[by['release_year'] >= 2000, :]
data = by.groupby('release_year').sum().reset_index()
a1 = alt.Chart(data).mark_circle(color='red').encode(
    x='release_year:O',
    y='budget:Q',
    size='budget:Q',
    tooltip=['release_year', 'budget']
).properties(
    title='How budget/revenue varies each year'
)
a2 = alt.Chart(data).mark_circle(opacity=0.3, color='green').encode(
    x='release_year:O',
    y='revenue:Q',
    size='revenue:Q',
    tooltip=['release_year', 'revenue']
)
alt.layer(a1, a2)

In [ ]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

# Asking and Answering Questions

## Most succesful actors based on revenue?.

In [ ]:
actors = movies.copy()
actors.set_index('id', inplace=True)
actor = actors.cast.str.split("|", expand = True)
actor = actor.stack().reset_index(level=1, drop=True).to_frame()
actor.columns = ['actor']
actor = actor.merge(actors[['title', 'revenue', 'budget', 'vote_average', 'popularity']], 
            how='left', left_index=True, right_index=True)

### Most number of movies acted by.

In [ ]:
# Most number of movies acted by.
act = actor.actor.value_counts().reset_index()
act.columns = ['actor', 'counts']

bars = alt.Chart(act.head(20)).mark_bar().encode(
    x = 'actor:O',
    y = 'counts:Q',
    color = alt.condition(
        datum.counts == np.amax(act.counts),
        alt.value('red'),
        alt.value('green')
    ),
    tooltip='actor'
).properties(
    title='Most Number of Films Acted by',
    height=200,
    width=500
)
text = bars.mark_text(
    align='center', # center left right
    baseline='bottom', # top middle
    dx=0  # Nudges text to right so it doesn't appear on top of the bar
).encode(
    text='counts:Q'
)
alt.layer(bars, text)


### Actors with highest revenue

In [ ]:
# ['title', 'revenue', 'budget', 'vote_average', 'popularity', 'profit', 'return']
act = actor.groupby("actor").agg(Total_Revenue = ("revenue", "sum"), 
                                 Total_Budget = ("budget", 'sum'),
                               Mean_Rating = ("vote_average", "mean"), 
                               Mean_Popularity = ("popularity", "mean"), 
                               Total_Movies = ("title", "count"))
act = act.sort_values(by = ['Total_Revenue'], ascending = False)
# act.nlargest(10, 'Total_Revenue')
bar = alt.Chart(act.reset_index().head(20)).mark_circle(color='green').encode(
    x='actor:O',
    y=alt.Y('Total_Revenue:Q'),
    size='Total_Revenue:Q',
    tooltip=['actor', 'Total_Revenue'],
    color = alt.condition(
        datum.Total_Revenue == np.amax(act.reset_index().Total_Revenue),
        alt.value('purple'),
        alt.value('green'))
).properties(
    title='Actors with Highest Revenue'
)
area = alt.Chart(act.reset_index().head(20)).mark_point(color='red').encode(
    x='actor:O',
    y='Total_Budget:Q',
    size='Total_Budget:Q',
    tooltip=['actor', 'Total_Budget'],
    color = alt.condition(
        datum.Total_Budget == np.amax(act.reset_index().Total_Budget),
        alt.value('purple'),
        alt.value('red'))
)
alt.layer(bar, area)

In [ ]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

## Are film series succesful?.

In [ ]:
movies.belongs_to_collection.head()
# not nan values :- true and nan :- false
movies['series'] = movies.belongs_to_collection.notna()
# series = movies.copy()

In [ ]:
# films grouped by series
movies.groupby('series').aggregate({'title':'count', 'budget':['sum', 'mean'],
                                    'revenue':['sum', 'mean'], 'profit':'mean',
                                    'returns':'median', 'vote_count':'mean',
                                    'vote_average':'mean','popularity':'mean'})

In [ ]:
series = movies.groupby('belongs_to_collection').aggregate({'title':'count', 'budget':['sum', 'mean'],
                                    'revenue':['sum', 'mean'], 'profit':'mean',
                                    'returns':'mean', 'vote_count':'mean',
                                    'vote_average':'mean','popularity':'mean'})

### Which series has highest number of movies?.

In [ ]:
series.nlargest(10, ('title', 'count'))

### Which series has highest revenue?.

In [ ]:
series.nlargest(10, ('revenue', 'sum'))

### Which series has highest mean budget?.

In [ ]:
series.nlargest(10, ('budget', 'mean'))

### Which series has highest return?.

In [ ]:
ser = movies.dropna(subset=['budget'])
series = ser.groupby('belongs_to_collection').aggregate({'title':'count', 'budget':['mean'],
                                    'revenue':['sum', 'mean'], 'profit':'mean',
                                    'returns':'mean', 'vote_count':'mean',
                                    'vote_average':'mean','popularity':'mean'})
series.nlargest(10, ('returns', 'mean'))

In [ ]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

## Find best production companies?.

In [ ]:
movies.production_companies.value_counts().head(5)
pc = movies.copy()
pc.set_index('id', inplace=True)
company = pc.production_companies.str.split("|", expand = True)
company = company.stack().reset_index(level=1, drop=True).to_frame()
company.columns = ['production_companies']
company = company.merge(pc[['title', 'revenue', 'budget', 'vote_average', 'popularity', 'profit', 'returns']], 
            how='left', left_index=True, right_index=True)
company.drop_duplicates(subset=['production_companies', 'title'], inplace=True)

### Which company produced most movies?

In [ ]:
# company[company['production_companies'] == 'Warner Bros.'].title.value_counts(ascending=False)
company['production_companies'].value_counts().head(10)

### Companies with most revenue?

In [ ]:
company.groupby('production_companies').agg({'revenue':'sum', 'budget':'sum',
                                            'vote_average':'mean', 'popularity':'mean',
                                            'profit':'sum'}).sort_values('revenue', ascending=False).head(10)

### Companies most popular with budget spent above 5000 million usd?

In [ ]:
cb1000 = company.groupby('production_companies').agg({'revenue':'sum', 'budget':'sum',
                                            'vote_average':'mean', 'popularity':'mean',
                                            'profit':'sum'})
cb1000.loc[cb1000.budget > 5000].sort_values('popularity', ascending=False).head(10)

In [ ]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

## Find Hit and Flop movies based on votes, rating, popularity and budget?

In [ ]:
# function to find hit and flop movies
def hit_flop(n, by, ascending=False, bud_min = 0, avg_vot = 0, vot_min = 0, pop_min = 0):
    data = movies.loc[((movies.budget >= bud_min) & (movies.vote_average >= avg_vot) & (movies.vote_count >= vot_min) & (movies.popularity >= pop_min)), :]
    data = data.sort_values(by=by, ascending=ascending).head(n).copy()
    data = data.loc[:, ['title', 'director', 'screenplay', 'budget', 'revenue', 'profit', 'returns', 'vote_average', 'vote_count', 'popularity', 'production_companies']]
    data = data.reset_index(level=0, drop=True)
    return data

### Movies with high budget?

In [ ]:
hit_flop(5, by='budget')

### Highested rated movies?

In [ ]:
hit_flop(5, by='vote_average', bud_min=10, vot_min = 500, pop_min = 20)

### Hit movies

In [ ]:
hit_flop(5, by='returns', bud_min=10)

### Flop movies

In [ ]:
hit_flop(5, by='returns', bud_min=50, ascending=True)

In [ ]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

## Genres and movies success?.

In [ ]:
genres = movies.copy()
genres.set_index('id', inplace=True)
genre = genres['genres'].str.split("|", expand = True)
genre = genre.stack().reset_index(level=1, drop=True).to_frame()
genre.columns = ['genres']
genre = genre.merge(genres[['title', 'revenue', 'budget', 'vote_average', 'popularity', 'release_date']], 
            how='left', left_index=True, right_index=True)
genre.drop_duplicates(subset=['genres', 'title'], inplace=True)
genre['year'] = genre.release_date.apply(lambda x: x.year)
genre.drop(columns=['release_date'], inplace=True)
genre = genre.convert_dtypes()

### Which genre has most movies?

In [ ]:
genre.genres.value_counts().head()

### Genres popular year to year

In [ ]:
genre.sort_values(by='year', ascending=False, inplace=True)
gen = genre.loc[genre.year > 1999, :]
# gpyy = gen.groupby(['year', 'genres']).agg({'title':'count',
#                                     'revenue':'sum',
#                                      'budget':'sum',
#                                      'vote_average':'mean',
#                                      'popularity':'sum'
#                                     })
gpyy = gen.groupby(['year', 'genres']).agg(total_no=('title', 'count'),
                                    revenue = ('revenue', 'mean'),
                                    budget = ('budget', 'mean'),
                                    vote = ('vote_average', 'mean'),
                                    popularity = ('popularity', 'mean'))
                                    
col = ['revenue', 'budget', 'vote', 'popularity']
for value in col:
    gpyy[col] = gpyy[col].apply(lambda x:round(x, ndigits=2))

In [ ]:
gp = gpyy.reset_index().sort_values(by=['year','total_no'], ascending=[False, False])
gp.head()

In [ ]:
# highest released genre.
gpyy.nlargest(5, columns=['total_no'])

In [ ]:
# highest revenue generated genre all years
gpyy.nlargest(5, columns=['revenue'])

In [ ]:
# highest popular genre all years
gp.loc[gp.loc[:,'total_no'] > 10, :].sort_values(by=['popularity'], ascending=[False]).head()

In [ ]:
# genres popular each year
total_no = alt.Chart(gpyy.reset_index()).mark_area().encode(
    x='year:O',
    y='total_no:Q',
    color='genres:N',
    size='total_no:Q',
    tooltip=['total_no', 'genres', 'year']
).interactive().properties(title='Total number of different genres released each year.')
revenue = alt.Chart(gpyy.reset_index()).mark_area().encode(
    x='year:O',
    y='revenue:Q',
    color='genres:N',
    size='revenue:Q',
    tooltip=['revenue', 'genres', 'year']
).interactive().properties(title='Mean revenue of different genres released each year')
budget = alt.Chart(gpyy.reset_index()).mark_area().encode(
    x='year:O',
    y='budget:Q',
    color='genres:N',
    size='budget:Q',
    tooltip=['budget', 'genres', 'year']
).interactive().properties(title='Mean budget of different genres released each year')
vote = alt.Chart(gpyy.reset_index()).mark_area().encode(
    x='year:O',
    y='vote:Q',
    color='genres:N',
    size='vote:Q',
    tooltip=['vote', 'genres', 'year']
).interactive().properties(title='Mean vote of different genres released each year')
popularity = alt.Chart(gpyy.reset_index()).mark_area().encode(
    x='year:O',
    y='popularity:Q',
    color='genres:N',
    size='popularity:Q',
    tooltip=['popularity', 'genres', 'year']
).interactive().properties(title='Mean popularity of different genres released each year')
alt.hconcat(total_no, revenue, budget, popularity, vote)

In [ ]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)

# Inferences and Conclusions

* Movies directed by John Ford - 66, Michael Curtiz - 65 are highest.
* Most succesful director by profit are Steven Spielberg: with 7520 million usd, Peter Jackson: with 5202 million usd.  
* Movies written by Woody Allen - 32, Nunnally Johnson - 28 are highest.  
* Most succesful writers by profit are Steve Kloves: with 5636 million usd, James Cameron: with 4610 million usd.  
* Maximum profit is for movie Avatar by James Cameron, 2550 million usd.  
* Highest number of movies released is in 2014, 1961 movies.  
* Revenue shoot up linearly with year, while budget increases slightly every year.  
* Bess flowers acted in 240 movies is highest.
* Actors with highest total revenues are Stan Lee: 19414 million usd and Samuel L Jackson: 17109 million usd.  
* Film Series are very succesful, mean revenue, return are higher than non series and more popular than non series.  
* Largest series by revenue are Harry Potter series, Star Wars series, James Bond series.  
* Warner Bros and followed by MGM are top in movie production.  
* Highest revenue is obtained for Warner Bros and second is for Universal Pictures.  
* Most popular production companies are Village Roadshow and Dreamworks SKG.  
* Pirates of the Caribbean. On Stranger Tides (380 musd) and Pirates of the Caribbean. At's world's end (300 musd) are the highest budget movies.  
* Dilwale Dulhania Le Jayenge and The Shawshank Redemption are highest rated movies.  
* E.T. The Extra Terrestrial and Star Wars has highest return.  
* Drama, followed by Comedy and Thriller movies are produced more.  
* Adventure, followed by Science Fiction and Action are most popular genres.


In [ ]:
# commit to jovian.ml
jovian.commit(project='my-project', environment=None)